File: apps/backend-java/src/main/java/com/pawpals/backend/post/PostsService.java
package com.pawpals.backend.post;

import com.pawpals.backend.common.ApiException;
import com.pawpals.backend.config.AppProperties;
import com.pawpals.backend.discover.dto.MeResponse;
import com.pawpals.backend.pet.Pet;
import com.pawpals.backend.pet.PetRepository;
import com.pawpals.backend.post.dto.AuthorSummary;
import com.pawpals.backend.post.dto.CreatePostRequest;
import com.pawpals.backend.post.dto.PostCommentResponse;
import com.pawpals.backend.post.dto.PostPhotoResponse;
import com.pawpals.backend.post.dto.PostResponse;
import com.pawpals.backend.post.dto.TipRequest;
import com.pawpals.backend.post.dto.TipperEntry;
import com.pawpals.backend.user.User;
import com.pawpals.backend.user.UserRepository;
import com.pawpals.backend.wallet.WalletRepository;
import com.pawpals.backend.wallet.WalletTransaction;
import com.pawpals.backend.wallet.WalletTransactionRepository;
import com.pawpals.backend.wallet.WalletTransactionType;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.List;
import java.util.Set;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

@Service
@RequiredArgsConstructor
public class PostsService {

  private static final int MAX_PHOTOS_PER_POST = 5;
  private static final Set<String> ALLOWED_TYPES = Set.of("image/jpeg", "image/png");

  private final PostRepository postRepository;
  private final PostPhotoRepository postPhotoRepository;
  private final PostLikeRepository postLikeRepository;
  private final PostCommentRepository postCommentRepository;
  private final PostTipRepository postTipRepository;
  private final PetRepository petRepository;
  private final UserRepository userRepository;
  private final WalletRepository walletRepository;
  private final WalletTransactionRepository walletTransactionRepository;
  private final AppProperties appProperties;

  public PostResponse create(String authorId, CreatePostRequest dto) {
    Pet pet = null;
    if (dto.petId() != null) {
      pet = petRepository.findById(dto.petId()).orElse(null);
      if (pet == null || !pet.getOwnerId().equals(authorId)) {
        throw ApiException.forbidden("PET_NOT_OWNED");
      }
    }
    Post post = new Post();
    post.setAuthor(userRepository.getReferenceById(authorId));
    post.setPet(pet);
    post.setCaption(dto.caption());
    post = postRepository.save(post);
    return toResponse(post, List.of(), null);
  }

  public PostPhotoResponse addPhoto(String postId, String userId, MultipartFile file, int sortOrder) {
    Post post = postRepository.findById(postId).orElseThrow(() -> ApiException.notFound("POST_NOT_FOUND"));
    if (!post.getAuthorId().equals(userId)) {
      throw ApiException.forbidden("FORBIDDEN");
    }
    if (postPhotoRepository.countByPost_Id(postId) >= MAX_PHOTOS_PER_POST) {
      throw ApiException.unprocessable("MAX_PHOTOS_REACHED");
    }
    if (file == null || file.isEmpty()) {
      throw ApiException.unprocessable("FILE_REQUIRED");
    }
    if (!ALLOWED_TYPES.contains(file.getContentType())) {
      throw ApiException.unprocessable("INVALID_FILE_TYPE");
    }

    String ext = extensionOf(file.getOriginalFilename());
    String filename = System.currentTimeMillis() + ext;
    Path dir = Path.of(appProperties.uploads().dir(), "posts", postId).toAbsolutePath();
    try {
      Files.createDirectories(dir);
      file.transferTo(dir.resolve(filename));
    } catch (IOException e) {
      throw new UncheckedIOException(e);
    }

    PostPhoto photo = new PostPhoto();
    photo.setPost(post);
    photo.setUrl("/uploads/posts/" + postId + "/" + filename);
    photo.setSortOrder(sortOrder);
    return PostPhotoResponse.from(postPhotoRepository.save(photo));
  }

  public void delete(String postId, String userId) {
    Post post = postRepository.findById(postId).orElseThrow(() -> ApiException.notFound("POST_NOT_FOUND"));
    if (!post.getAuthorId().equals(userId)) {
      throw ApiException.forbidden("FORBIDDEN");
    }
    postRepository.delete(post);
  }

  /** 配對前的探索層：只顯示附近使用者的貼文，不是全域動態牆。 */
  public List<PostResponse> getFeed(String userId, int radiusKm, int page, int limit) {
    User me = userRepository.findById(userId).orElseThrow(() -> ApiException.notFound("USER_NOT_FOUND"));
    if (me.getLatitude() == null || me.getLongitude() == null) {
      return List.of();
    }

    double radiusM = radiusKm * 1000.0;
    List<Post> posts = postRepository.findAllByOrderByCreatedAtDesc();

    record Scored(Post post, User author, double distanceM) {}

    List<Scored> withDistance =
        posts.stream()
            .map(p -> {
              User author = userRepository.findById(p.getAuthorId()).orElse(null);
              if (author == null || author.getLatitude() == null || author.getLongitude() == null) {
                return null;
              }
              double d =
                  haversineM(me.getLatitude(), me.getLongitude(), author.getLatitude(), author.getLongitude());
              return new Scored(p, author, d);
            })
            .filter(s -> s != null && s.distanceM() <= radiusM)
            .toList();

    int offset = Math.max(page - 1, 0) * limit;
    return withDistance.stream()
        .skip(offset)
        .limit(limit)
        .map(s -> toResponse(s.post(), photosOf(s.post().getId()), s.distanceM(), s.author()))
        .toList();
  }

  public PostResponse getOne(String postId) {
    Post post = postRepository.findById(postId).orElseThrow(() -> ApiException.notFound("POST_NOT_FOUND"));
    return toResponse(post, photosOf(postId), null);
  }

  public void like(String postId, String userId) {
    ensurePostExists(postId);
    if (postLikeRepository.findByPost_IdAndUser_Id(postId, userId).isEmpty()) {
      PostLike like = new PostLike();
      like.setPost(postRepository.getReferenceById(postId));
      like.setUser(userRepository.getReferenceById(userId));
      postLikeRepository.save(like);
    }
  }

  @Transactional
  public void unlike(String postId, String userId) {
    postLikeRepository.deleteByPost_IdAndUser_Id(postId, userId);
  }

  public PostCommentResponse addComment(String postId, String userId, String text) {
    ensurePostExists(postId);
    PostComment comment = new PostComment();
    comment.setPost(postRepository.getReferenceById(postId));
    comment.setAuthor(userRepository.getReferenceById(userId));
    comment.setText(text);
    return PostCommentResponse.from(postCommentRepository.save(comment));
  }

  public List<PostCommentResponse> getComments(String postId, int limit) {
    return postCommentRepository.findByPost_IdOrderByCreatedAtDesc(postId, PageRequest.of(0, limit)).stream()
        .map(PostCommentResponse::from)
        .toList();
  }

  /**
   * 抖內：肉乾只能花在「打賞貼文」這個定義好的用途，不能任意轉帳給其他使用者。
   * 唯一的額外效果是把贈送者加進 PostTip，讓貼文作者之後能查看贈送者的個人介紹（見 getTippers）。
   */
  @Transactional
  public String tip(String postId, String senderId, TipRequest dto) {
    if (Arrays.stream(TipRequest.ALLOWED_AMOUNTS).noneMatch(a -> a == dto.amount())) {
      throw ApiException.unprocessable("INVALID_AMOUNT");
    }
    Post post = postRepository.findById(postId).orElseThrow(() -> ApiException.notFound("POST_NOT_FOUND"));
    if (post.getAuthorId().equals(senderId)) {
      throw ApiException.forbidden("CANNOT_TIP_OWN_POST");
    }
    int amount = dto.amount();

    var senderWallet = walletRepository.lockByUserId(senderId).orElseThrow(() -> ApiException.notFound("WALLET_NOT_FOUND"));
    if (senderWallet.getBalance() < amount) {
      throw ApiException.unprocessable("INSUFFICIENT_BALANCE");
    }
    var recipientWallet =
        walletRepository.lockByUserId(post.getAuthorId()).orElseThrow(() -> ApiException.notFound("WALLET_NOT_FOUND"));

    senderWallet.setBalance(senderWallet.getBalance() - amount);
    recipientWallet.setBalance(recipientWallet.getBalance() + amount);
    walletRepository.save(senderWallet);
    walletRepository.save(recipientWallet);

    recordTransaction(senderId, WalletTransactionType.TIP_SENT, amount, postId);
    recordTransaction(post.getAuthorId(), WalletTransactionType.TIP_RECEIVED, amount, postId);

    PostTip tip = new PostTip();
    tip.setPost(post);
    tip.setSender(userRepository.getReferenceById(senderId));
    tip.setRecipient(post.getAuthor());
    tip.setAmount(amount);
    tip = postTipRepository.save(tip);
    return tip.getId();
  }

  /** 只有貼文作者可以看到「誰抖內了、他們的個人介紹長怎樣」。 */
  public List<TipperEntry> getTippers(String postId, String requesterId) {
    Post post = postRepository.findById(postId).orElseThrow(() -> ApiException.notFound("POST_NOT_FOUND"));
    if (!post.getAuthorId().equals(requesterId)) {
      throw ApiException.forbidden("FORBIDDEN");
    }
    return postTipRepository.findByPost_IdOrderByCreatedAtDesc(postId).stream()
        .map(t -> {
          User sender = userRepository.findById(t.getSenderId()).orElseThrow(() -> ApiException.notFound("USER_NOT_FOUND"));
          return new TipperEntry(t.getAmount(), t.getCreatedAt(), MeResponse.from(sender));
        })
        .toList();
  }

  private void ensurePostExists(String postId) {
    if (!postRepository.existsById(postId)) {
      throw ApiException.notFound("POST_NOT_FOUND");
    }
  }

  private void recordTransaction(String userId, WalletTransactionType type, int amount, String relatedEntityId) {
    WalletTransaction tx = new WalletTransaction();
    tx.setUser(userRepository.getReferenceById(userId));
    tx.setType(type);
    tx.setAmount(amount);
    tx.setRelatedEntityId(relatedEntityId);
    walletTransactionRepository.save(tx);
  }

  private List<PostPhotoResponse> photosOf(String postId) {
    return postPhotoRepository.findByPost_IdOrderBySortOrderAsc(postId).stream().map(PostPhotoResponse::from).toList();
  }

  private PostResponse toResponse(Post post, List<PostPhotoResponse> photos, Double distanceM) {
    User author = userRepository.findById(post.getAuthorId()).orElse(null);
    return toResponse(post, photos, distanceM, author);
  }

  private PostResponse toResponse(Post post, List<PostPhotoResponse> photos, Double distanceM, User author) {
    long likeCount = postLikeRepository.countByPost_Id(post.getId());
    long commentCount = postCommentRepository.countByPost_Id(post.getId());
    long tipCount = postTipRepository.countByPost_Id(post.getId());
    AuthorSummary authorSummary = author != null ? AuthorSummary.from(author) : null;
    return new PostResponse(
        post.getId(),
        post.getAuthorId(),
        post.getPetId(),
        post.getCaption(),
        post.getCreatedAt(),
        photos,
        likeCount,
        commentCount,
        tipCount,
        authorSummary,
        distanceM);
  }

  private double haversineM(double lat1, double lon1, double lat2, double lon2) {
    double R = 6371000;
    double dLat = Math.toRadians(lat2 - lat1);
    double dLon = Math.toRadians(lon2 - lon1);
    double a =
        Math.pow(Math.sin(dLat / 2), 2)
            + Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2)) * Math.pow(Math.sin(dLon / 2), 2);
    return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  }

  private String extensionOf(String originalFilename) {
    if (originalFilename == null) return "";
    int dot = originalFilename.lastIndexOf('.');
    return dot >= 0 ? originalFilename.substring(dot) : "";
  }
}

File: apps/backend/src/posts/posts.service.ts
import {
  BadRequestException,
  ForbiddenException,
  Injectable,
  NotFoundException,
  UnprocessableEntityException,
} from '@nestjs/common';
import { Prisma } from '@prisma/client';
import { PrismaService } from '../prisma/prisma.service';
import { StorageService } from '../storage/storage.service';
import { CreatePostDto } from './dto/create-post.dto';
import { extname } from 'path';

const MAX_PHOTOS_PER_POST = 5;

// 跟 discover.controller.ts 的 USER_SELECT 同一組欄位：這就是抖內解鎖後，貼文作者看到的「個人介紹」。
const TIPPER_SELECT = {
  id: true, email: true, role: true, gender: true, displayName: true,
  bio: true, interests: true, education: true, zodiac: true,
  jobTitle: true, company: true, school: true, city: true,
  height: true, avatarUrl: true, createdAt: true,
};

const POST_AUTHOR_SELECT = {
  id: true, displayName: true, avatarUrl: true, city: true,
};

function haversineM(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const R = 6371000;
  const dLat = (lat2 - lat1) * Math.PI / 180;
  const dLon = (lon2 - lon1) * Math.PI / 180;
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) * Math.sin(dLon / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

@Injectable()
export class PostsService {
  constructor(private prisma: PrismaService, private storageService: StorageService) {}

  async create(authorId: string, dto: CreatePostDto) {
    if (dto.petId) {
      const pet = await this.prisma.pet.findUnique({ where: { id: dto.petId } });
      if (!pet || pet.ownerId !== authorId) throw new ForbiddenException('PET_NOT_OWNED');
    }
    return this.prisma.post.create({
      data: { authorId, petId: dto.petId, caption: dto.caption },
      include: { photos: true },
    });
  }

  async addPhoto(postId: string, userId: string, file: Express.Multer.File, sortOrder: number) {
    const post = await this.prisma.post.findUnique({ where: { id: postId }, include: { photos: true } });
    if (!post) throw new NotFoundException('POST_NOT_FOUND');
    if (post.authorId !== userId) throw new ForbiddenException();
    if (post.photos.length >= MAX_PHOTOS_PER_POST) throw new UnprocessableEntityException('MAX_PHOTOS_REACHED');
    if (!file) throw new BadRequestException('FILE_REQUIRED');

    const key = `posts/${postId}/${Date.now()}${extname(file.originalname)}`;
    const url = await this.storageService.uploadBuffer(file.buffer, key, file.mimetype);
    return this.prisma.postPhoto.create({ data: { postId, url, sortOrder } });
  }

  async delete(postId: string, userId: string) {
    const post = await this.prisma.post.findUnique({ where: { id: postId } });
    if (!post) throw new NotFoundException('POST_NOT_FOUND');
    if (post.authorId !== userId) throw new ForbiddenException();
    await this.prisma.post.delete({ where: { id: postId } });
    return { ok: true };
  }

  /** 配對前的探索層：只顯示附近使用者的貼文，不是全域動態牆。 */
  async getFeed(userId: string, radiusKm = 5, page = 1, limit = 20) {
    const me = await this.prisma.user.findUnique({
      where: { id: userId },
      select: { latitude: true, longitude: true },
    });
    if (!me?.latitude || !me?.longitude) return [];

    const posts = await this.prisma.post.findMany({
      where: { author: { latitude: { not: null }, longitude: { not: null } } },
      include: {
        author: { select: { ...POST_AUTHOR_SELECT, latitude: true, longitude: true } },
        pet: { select: { id: true, name: true, breed: true } },
        photos: { orderBy: { sortOrder: 'asc' } },
        _count: { select: { likes: true, comments: true, tips: true } },
      },
      orderBy: { createdAt: 'desc' },
    });

    const radiusM = radiusKm * 1000;
    const withDistance = posts
      .map((p) => ({
        post: p,
        distanceM: haversineM(me.latitude!, me.longitude!, p.author.latitude!, p.author.longitude!),
      }))
      .filter((p) => p.distanceM <= radiusM);

    const offset = (page - 1) * limit;
    return withDistance.slice(offset, offset + limit).map(({ post, distanceM }) => {
      const { author, ...rest } = post;
      const { latitude, longitude, ...authorRest } = author;
      return { ...rest, author: authorRest, distanceM };
    });
  }

  async getOne(postId: string) {
    const post = await this.prisma.post.findUnique({
      where: { id: postId },
      include: {
        author: { select: POST_AUTHOR_SELECT },
        pet: { select: { id: true, name: true, breed: true } },
        photos: { orderBy: { sortOrder: 'asc' } },
        _count: { select: { likes: true, comments: true, tips: true } },
      },
    });
    if (!post) throw new NotFoundException('POST_NOT_FOUND');
    return post;
  }

  async like(postId: string, userId: string) {
    await this.ensurePostExists(postId);
    await this.prisma.postLike.upsert({
      where: { postId_userId: { postId, userId } },
      create: { postId, userId },
      update: {},
    });
    return { ok: true };
  }

  async unlike(postId: string, userId: string) {
    await this.prisma.postLike.deleteMany({ where: { postId, userId } });
    return { ok: true };
  }

  async addComment(postId: string, userId: string, text: string) {
    await this.ensurePostExists(postId);
    return this.prisma.postComment.create({
      data: { postId, authorId: userId, text },
    });
  }

  async getComments(postId: string, limit = 20) {
    return this.prisma.postComment.findMany({
      where: { postId },
      orderBy: { createdAt: 'desc' },
      take: limit,
      include: { author: { select: POST_AUTHOR_SELECT } },
    });
  }

  /**
   * 抖內：肉乾只能花在「打賞貼文」這個定義好的用途，不能任意轉帳給其他使用者。
   * 唯一的額外效果是把贈送者加進 PostTip，讓貼文作者之後能查看贈送者的個人介紹（見 getTippers）。
   */
  async tip(postId: string, senderId: string, amount: number) {
    const post = await this.prisma.post.findUnique({ where: { id: postId } });
    if (!post) throw new NotFoundException('POST_NOT_FOUND');
    if (post.authorId === senderId) throw new ForbiddenException('CANNOT_TIP_OWN_POST');

    return this.prisma.$transaction(async (tx: Prisma.TransactionClient) => {
      const senderWallet = await tx.wallet.findUnique({ where: { userId: senderId } });
      if (!senderWallet || senderWallet.balance < amount) {
        throw new UnprocessableEntityException('INSUFFICIENT_BALANCE');
      }

      await tx.wallet.update({ where: { userId: senderId }, data: { balance: { decrement: amount } } });
      await tx.wallet.update({ where: { userId: post.authorId }, data: { balance: { increment: amount } } });
      await tx.walletTransaction.createMany({
        data: [
          { userId: senderId, type: 'TIP_SENT', amount, relatedEntityId: postId },
          { userId: post.authorId, type: 'TIP_RECEIVED', amount, relatedEntityId: postId },
        ],
      });
      const tip = await tx.postTip.create({
        data: { postId, senderId, recipientId: post.authorId, amount },
      });
      return { ok: true, tipId: tip.id };
    });
  }

  /** 只有貼文作者可以看到「誰抖內了、他們的個人介紹長怎樣」。 */
  async getTippers(postId: string, requesterId: string) {
    const post = await this.prisma.post.findUnique({ where: { id: postId } });
    if (!post) throw new NotFoundException('POST_NOT_FOUND');
    if (post.authorId !== requesterId) throw new ForbiddenException();

    const tips = await this.prisma.postTip.findMany({
      where: { postId },
      orderBy: { createdAt: 'desc' },
      include: { sender: { select: TIPPER_SELECT } },
    });
    return tips.map((t) => ({ amount: t.amount, tippedAt: t.createdAt, sender: t.sender }));
  }

  private async ensurePostExists(postId: string) {
    const exists = await this.prisma.post.findUnique({ where: { id: postId }, select: { id: true } });
    if (!exists) throw new NotFoundException('POST_NOT_FOUND');
  }
}
